examples/gradients.rs

```diff
@@ -4,29 +4,29 @@
 //!
 //! `cargo run --example gradients ./examples/empire-state-building.jpg ./tmp`
 
-use image::{open, GrayImage, Luma};
-use imageproc::{
-    definitions::Image,
-    gradients::{
-        horizontal_prewitt, horizontal_scharr, horizontal_sobel, vertical_prewitt, vertical_scharr,
-        vertical_sobel,
-    },
-    map::map_pixels,
-};
+use image::{open, GrayImage};
+use imageproc::{filter::filter3x3, gradients::GradientKernel, map::map_subpixels};
 use std::{env, fs, path::Path};
 
-fn compute_gradients<F>(
+fn save_gradients(
     input: &GrayImage,
-    gradient_func: F,
+    gradient_kernel: &GradientKernel,
     output_dir: &Path,
-    file_name: &str,
+    name: &str,
     scale: i16,
-) where
-    F: Fn(&GrayImage) -> Image<Luma<i16>>,
-{
-    let gradient = gradient_func(input);
-    let gradient = map_pixels(&gradient, |_, _, p| Luma([(p[0].abs() / scale) as u8]));
-    gradient.save(&output_dir.join(file_name)).unwrap();
+) {
+    let horizontal_gradients = filter3x3(input, gradient_kernel.kernel1::<i16>());
```

Comment:
I don't know why should change filter3x3 signature. Just `Kernel::filter`

Reply:
I don't know what you mean?

Reply:
The reason is because of #520 where users are struggling to discover the `filter()` method because it is a associated method rather than a free-standing function like every other function in the library. This is very inconsistent and so it would be better to just use `filter3x3(image, kernel)` or `filter(image, kernel)` instead of `kernel.filter(image)`.

Reply:
I think we can do it without breaking changes!
What about `pub fn filter(image: &Image<P>, kernel: &Kernel<K>, f: impl FnMut(&mut Q::Subpixel, K)) -> Image<Q>`?
We can also implement `Deref`, so we can call `filter3x3(&img, &kernel)`
```
impl<'a, T> Deref for Kernel<'a, T> {
    type Target = [T];
    fn deref(&self) -> &Self::Target {
        self.data
    }
}
```

Reply:
It's not difficult to find the `Kernel` methods, we just don't have examples. Also there is no `clamped_filter` method, which is easier to use

Reply:
I added a `filter_clamped()` method in this PR for convenience.

> pub fn filter<'a>(image: &Image<P>, kernel: &Kernel<'a, K>, f: impl FnMut(&mut Q::Subpixel, K)) -> Image<Q>

This is already possible by the `filter()` method in this PR

Reply:
But you have trait

Reply:
Yes the trait is used in the `filter_clamped()` function. Its signature is simply: `pub fn filter_clamped<P, K, S>(image: &Image<P>, kernel: impl Kernel<K>) -> Image<ChannelMap<P, S>>`

Reply:
`OwnedKernel` and `BorrowedKernel` are the main two types which now implement the `Kernel` trait.